File: slp/backend-dotnet/Features/Queue/NullQueueService.cs
namespace backend_dotnet.Features.Queue;

/// <summary>
/// No-op implementation of <see cref="IQueueService"/>.
/// Registered when <c>Queue:Enabled</c> is <c>false</c>;
/// the controller falls through to synchronous processing in that case
/// and never calls these methods.
/// </summary>
public class NullQueueService : IQueueService
{
    public bool IsAvailable => false;

    public Task EnqueueAsync(LlmJob job) =>
        throw new InvalidOperationException(
            "Queue is disabled. Set Queue:Enabled = true in configuration to use async processing.");

    public Task<LlmJob?> DequeueAsync(CancellationToken ct = default) =>
        Task.FromResult<LlmJob?>(null);

    public Task AcknowledgeAsync(string jobId) => Task.CompletedTask;

    public Task<List<string>> GetProcessingJobIdsAsync() =>
        Task.FromResult(new List<string>());

    public Task RequeueStaleAsync(string jobId) => Task.CompletedTask;
}

File: slp/backend-nestjs/src/modules/queue/null-queue.service.ts
import { Injectable } from '@nestjs/common';
import { IQueueService } from './queue.service.interface';
import { LlmJob } from './llm-job.interface';

@Injectable()
export class NullQueueService implements IQueueService {
  isAvailable = false;

  enqueue(_job: LlmJob): Promise<void> {
    throw new Error('Queue is disabled. Enable Queue:Enabled in configuration.');
  }

  dequeue(): Promise<LlmJob | null> {
    return Promise.resolve(null);
  }

  acknowledge(_jobId: string): Promise<void> {
    return Promise.resolve();
  }

  getProcessingJobIds(): Promise<string[]> {
    return Promise.resolve([]);
  }

  requeueStale(_jobId: string): Promise<void> {
    return Promise.resolve();
  }
}